File: agent/skyhook-agent/src/skyhook_agent/interrupts.py
#  SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#  SPDX-License-Identifier: Apache-2.0
import string
import base64
import json
import inspect
import sys

class Interrupt(object):
    data: dict[str, any] = {}

    # These are the command(s) to be run to perform the interrupt. 
    # It is a list of subprocess.run list form.
    # Set this to empty list if you do NOT want anything run.

    interrupt_cmd: list[list[str]]

    @classmethod
    def _type(cls):
        return ''.join(c.lower() if c in string.ascii_lowercase or i == 0 else f"_{c.lower()}" for i,c in enumerate(cls.__name__))

    @property
    def type(self):
        return self._type()

    def serialize(self) -> dict[str, str|dict[str,str]]:
        return dict(type=self.type, **self.data)
    
    def make_controller_input(self) -> str:
        """
        Available as an example of the expected input to the inflate method
        """
        return base64.b64encode(str.encode(json.dumps(self.serialize()), 'utf-8'))

class NodeRestart(Interrupt):
    """
    Operator will do the interrupt for you by restarting the node.
    """
    
    interrupt_cmd = [["reboot",]]

class ServiceRestart(Interrupt):
    """
    Operator will do the interrupt for you by restarting the given service.
    """

    def __init__(self, services: list[str]):
        self.data = {
            "services": services
        }

        self.interrupt_cmd = [
            ["systemctl", "daemon-reload"]
        ]
        self.interrupt_cmd += [["systemctl", "restart", s] for s in services]

class RestartAllServices(Interrupt):
    """
    Operator will do the interrupt for you by restarting all services
    """

    interrupt_cmd = [
        ["service" ,"procps", "force-reload"]
    ]

class NoOp(Interrupt):
    """
    Operator will do nothing
    """
    interrupt_cmd = []

class ScriptInterrupt(Interrupt):
    """
    The apply script will do the interrupt.
    """
    interrupt_cmd = []

### DO NOT PUT INTERRUPTS BELOW THIS

def _make_map():
    """
    Generate a map of all classes above this to use with inflate.
    """
    interrupt_map = {}
    for _, member in inspect.getmembers(sys.modules[__name__], lambda x: inspect.isclass(x) and issubclass(x, Interrupt) and x != Interrupt):
        interrupt_map[member._type()] = member

    return interrupt_map


def inflate(serialized_value: str) -> Interrupt:
    """
    Convert base64 encoded Interrupt.serialize() back into its original 
    Interrupt.
    """
    interrupt_map = _make_map()

    try:
        serialized_data = json.loads(base64.b64decode(serialized_value))
    except:
        raise ValueError("Serialized interrupt must be base64 encoded {'type': str, **kwargs: dict[str, any]}")

    try:
        interrupt_type = serialized_data['type']
    except KeyError:
        raise ValueError("Serialized interrupt must be base64 encoded {'type': str, **kwargs: dict[str, any]}")

    interrupt_data = dict((k,v) for k, v in serialized_data.items() if k != "type")
    try:
        interrupt = interrupt_map[interrupt_type](**interrupt_data)
    except KeyError:
        raise ValueError(f"Unknown interrupt {interrupt_type} must be one of: {', '.join(interrupt_map.keys())}")
    
    return interrupt

File: agent/go/internal/interrupts/interrupts.go
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package interrupts decodes, encodes, and runs agent interrupt operations.
//
// Each Interrupt identifies its wire type with Type, executes its own command
// composition through Run using execution.Config and execution.Status, and
// preserves its operator-facing wire representation through Serialize.
package interrupts

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/NVIDIA/nodewright/agent/internal/execution"
)

// errInvalidSerializedInterrupt is returned when a serialized interrupt is not
// a base64-encoded JSON object of the expected shape. It is a sentinel so
// callers can branch on it with errors.Is.
var errInvalidSerializedInterrupt = errors.New(`serialized interrupt must be base64-encoded JSON with a "type" field`)

// InterruptType identifies a supported interrupt implementation on the wire.
type InterruptType string

const (
	NodeRestartType        InterruptType = "node_restart"
	ServiceRestartType     InterruptType = "service_restart"
	RestartAllServicesType InterruptType = "restart_all_services"
	NoOpType               InterruptType = "no_op"
	ScriptInterruptType    InterruptType = "script_interrupt"
)

// Interrupt is the contract satisfied by every interrupt type.
type Interrupt interface {
	Type() InterruptType
	Run(context.Context, execution.Config) (execution.Status, error)
	Serialize() ([]byte, error)
}

func validateRun(ctx context.Context, config execution.Config, interruptType InterruptType) error {
	if ctx == nil {
		return errors.New("running interrupt: context is nil")
	}
	if err := ctx.Err(); err != nil {
		return fmt.Errorf("running interrupt %q: %w", interruptType, err)
	}
	if err := config.Validate(); err != nil {
		return fmt.Errorf("running interrupt %q: invalid run config: %w", interruptType, err)
	}
	return nil
}

// Encode serializes an Interrupt to the base64+JSON wire form expected
// by the operator's interrupt controller input. A nil receiver is
// rejected explicitly so a misuse from a caller returns a typed error
// rather than panicking on a nil interface method call.
func Encode(i Interrupt) (string, error) {
	if i == nil {
		return "", errors.New("nil interrupt")
	}
	data, err := i.Serialize()
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(data), nil
}

// Decode parses a base64+JSON serialized interrupt produced by Encode
// (or its Python counterpart skyhook_agent.interrupts.inflate).
// Wire-shape errors return errInvalidSerializedInterrupt; unknown type
// names return a descriptive error listing the supported types.
func Decode(serializedValue string) (Interrupt, error) {
	data, err := base64.StdEncoding.DecodeString(serializedValue)
	if err != nil {
		return nil, fmt.Errorf("%w: %v", errInvalidSerializedInterrupt, err)
	}

	// type is probed as *string to distinguish an absent field (a wire-shape
	// error) from a present-but-empty value (an unknown-type error); the two
	// take different branches below.
	var head struct {
		Type *InterruptType `json:"type"`
	}
	if err := json.Unmarshal(data, &head); err != nil {
		return nil, fmt.Errorf("%w: %v", errInvalidSerializedInterrupt, err)
	}
	if head.Type == nil {
		return nil, errInvalidSerializedInterrupt
	}

	switch *head.Type {
	case NodeRestartType:
		return NodeRestart{}, nil
	case ServiceRestartType:
		var wire struct {
			Services []string `json:"services"`
		}
		if err := json.Unmarshal(data, &wire); err != nil {
			return nil, fmt.Errorf("%w: %v", errInvalidSerializedInterrupt, err)
		}
		return ServiceRestart{Services: wire.Services}, nil
	case RestartAllServicesType:
		return RestartAllServices{}, nil
	case NoOpType:
		return NoOp{}, nil
	case ScriptInterruptType:
		return ScriptInterrupt{}, nil
	default:
		return nil, fmt.Errorf(
			"unknown interrupt %q must be one of: %s",
			*head.Type,
			supportedTypes(),
		)
	}
}

// supportedTypes returns the known interrupt types in Decode dispatch order.
func supportedTypes() string {
	return strings.Join([]string{
		string(NodeRestartType),
		string(ServiceRestartType),
		string(RestartAllServicesType),
		string(NoOpType),
		string(ScriptInterruptType),
	}, ", ")
}

// marshalTypeOnly serializes an interrupt whose wire form is just
// {"type": "..."}.
func marshalTypeOnly(typeName InterruptType) ([]byte, error) {
	return json.Marshal(struct {
		Type InterruptType `json:"type"`
	}{Type: typeName})
}
